### ISSUE TITLE
Alternative implementations of an algorithm

### ISSUE THREAD

**Issue by lostatseajoshua** [raywenderlich/swift-algorithm-club#177]
## What is the issue?
Open discussion adding multiple methods of code to accomplish the same algorithm. There is always different ways to do the same thing. If it completes the same algorithm in the same time complexities it would be great to show different ways to implement. Some may understand different ways better.

## Example
This is proposing to add a quick sort and binary search alternative code examples to the one in the repo currently. The quick sort currently is a little more advanced. This proposes to add a simpler quick sort, no need an extra function for partition.

Quick sort alternative
``` swift
func quickSort<T: Comparable>(inout array: [T], left: Int, right: Int) -> [T] {
    if array.count <= 1 { return array }
    var i = left
    var j = right
    
    let mid = array[(left + right) / 2]
    
    repeat {
        while array[i] < mid && i < right {
            i += 1
        }

        while mid < array[j] && j > left {
            j -= 1
        }
        
        if i <= j {
            if i != j {
                swap(&array[i], &array[j])
            }
            i += 1
            j -= 1
        }
    } while (i <= j)
    
    if left < j {
        quickSort(&array, left: left, right: j)
    }
    
    if i < right {
        quickSort(&array, left: i, right: right)
    }
    
    return array
}
```

Binary search alternative

```swift
func binarySearch<T :Comparable>(arry: [T], key: T, low: Int, high: Int) -> Int? {
    if arry.count == 0 {
        return nil
    }
    
    if low > high {
        return nil
    }
    
    let mid = (low + high) / 2
    
    if arry[mid] == key {
        return mid
    }
    
    if arry[mid] < key {
        return binarySearch(arry, key: key, low: mid + 1, high: high)
    } else {
        return binarySearch(arry, key: key, low: low, high: mid - 1)
    }
}
```

### Discussion
This isn't about adding 15 ways to implement some simple algorithms but more about showing simpler ways to implement harder to comprehend algorithms such as merge sort, quick sort and binary tree. The more lower level the better. The algorithms should be written in Swift but not depend on Swift helper methods to achieve the goals. It defeats the purpose of understanding the underlying algorithm.

**Comment by kelvinlauKL** [comment]
I'm personally aware of this as I write the SAC posts at www.raywenderlich.com, but IMO I prefer to go as "Swifty" as possible. This generally means using methods specific to the Swift language from the standard library.

I've wrote 2 articles so far for the monthly algorithm articles at www.raywenderlich.com, and I did remember providing a lower level alternative to one of the methods that required the usage of `map`.

It'll be interesting to see how others feel about this.
